File: services/graph4j_service/src/process/src/main/java/process/algorithm/Algorithm.java
package process.algorithm;

import com.jsoniter.any.Any;
import org.graph4j.Graph;
import process.instance.Instance;
import process.pubsub.Publisher;

import java.util.List;

public abstract class Algorithm {
    protected Graph graph;
    protected Any parameters;
    protected List<Publisher> publishers;

    protected Algorithm(Instance instance, List<Publisher> publishers) {
        this.graph = instance.graph();
        this.parameters = instance.parameters();
        this.publishers = publishers;
    }

    protected abstract String algorithm();

    public void run() {
        String result = algorithm();
        publishResult(result);
    }

    private void publishResult(String result) {
        for(Publisher publisher : publishers) {
            publisher.send(result);
        }
    }
}

File: services/networkx_service/src/process/algorithm/algorithm.py
import os
import sys
from threading import Thread

class Algorithm:
    def __init__(self, instance, publishers):
        self.graph = instance.graph
        self.parameters = instance.parameters
        self.publishers = publishers

        self.best_heuristic_score = sys.float_info.max
        self.best_result = sys.float_info.max

    def algorithm(self):
        raise NotImplementedError("Unimplemented method")

    def run(self):
        thread = Thread(target=self.algorithm)
        thread.start()
        thread.join(timeout=int(os.environ["ALGORITHM_TIMEOUT"]))

        self.__publish_result(self.best_result, self.best_heuristic_score)

    def __publish_result(self, result, heuristic_score):
        for publisher in self.publishers:
            publisher.send(
                {
                    "result": result,
                    "heuristic_score": heuristic_score
                }
            )
